commit 2f1074037228cf09814d4a86dd8895616c6b7282 (refs/changes/75/575/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-10-03T12:10:49-07:00 (3 years ago)
    
    joyent/node-cueball#33 claim() should double-check state of ConnectionFSM

diff --git a/lib/pool.js b/lib/pool.js
index 98a97ab..9bc4a0b 100644
--- a/lib/pool.js
+++ b/lib/pool.js
@@ -605,9 +605,19 @@ CueBallConnectionPool.prototype.claim = function (options, cb) {
 	Error.captureStackTrace(e);
 
 	/* If there are idle connections sitting around, take one. */
-	if (this.p_idleq.length > 0) {
+	while (this.p_idleq.length > 0) {
 		var fsm = this.p_idleq.shift();
 		delete (fsm.p_idleq_node);
+		/*
+		 * Since 'stateChanged' is emitted async from mooremachine,
+		 * things may be on the idle queue still but not actually idle.
+		 * If we find one, just rip it off the queue (which we've
+		 * already done) and try the next thing. The state mgmt
+		 * callback from addConnection will cope.
+		 */
+		if (!fsm.isInState('idle'))
+			continue;
+
 		fsm.claim(e.stack, function (err, hdl, conn) {
 			if (err) {
 				if (!done)
